lib/marked.js

```diff
@@ -7,8 +7,223 @@
 ;(function(root) {
 'use strict';
 
+/**
+ * Begin Marked
+ *
+ * TODO: None of our tests actually test a user just passing src?
+ *       `console.log(marked.receivedOnlySource(src, opt, callback));`
+ */
+
+/**
+ * @param  src string
+ * @param  opt (optional) Object
+ * @param  callback (optional) Object
+ *
+ * @return string
+ */
+function marked(src, opt, callback) {
+  // ^^ Can't we do something like marked(src, opt = {}, callback = function())
+
+  try {
+    marked.receivedValidSource(src);
+  } catch (e) {
+    throw e;
+  }
+
+  console.log(marked.receivedOnlySource(src, opt, callback));
+
+  // We essentially have two `marked`...this one conditional, and the one at the end...!!
+  if (marked.receivedCallback(opt, callback)) {
+    // TODO: This seems like an odd check, think there's a more readable way to
+    //       deal with optional params.
+    //
+    // We have a callback but not as the last param passed in.
+    if (!callback) {
+      callback = opt;
+      opt = null;
+    }
+
+    // If we don't have a callback honestly passed in, opt becomes null,
+    // then we set it to a merged of the defalts + opt, or an empty object??
+    opt = merge({}, marked.defaults, opt || {});
+
+    // Not sure what all is happening here.
+    var highlight = opt.highlight,
+        tokens,
+        pending,
+        i = 0;
+
+    // Run the lexer, if it fails; bail early by calling the callback.
+    // Would like to move this somewhere else. Make most of the returns toward the end; or,
+    // more obvious.
+    try {
+      tokens = Lexer.lex(src, opt)
+    } catch (e) {
+      return callback(e);
+    }
+
+    // Internal block for promise-type behavior??
+    var done = function(err) {
+      if (err) {
+        opt.highlight = highlight;
+        return callback(err);
+      }
+
+      var out;
+
+      try {
+        out = Parser.parse(tokens, opt);
+      } catch (e) {
+        err = e;
+      }
+
+      opt.highlight = highlight;
+
+      return err
+        ? callback(err)
+        : callback(null, out);
+    };
+
+    // Highlight option check, not set or less than three bail
+    //
+    // TODO: Don't some highlighter plugins allow for `js`, effectively being less than three
+    //       despite wanting to have the highlighter run??
+    if (!highlight || highlight.length < 3) {
+      return done();
+    }
+
+    delete opt.highlight;
+
+    pending = tokens.length;
+    if (!pending) return done();
+
+    for (; i < tokens.length; i++) {
+      (function(token) {
+        if (token.type !== 'code') {
+          return --pending || done();
+        }
+        return highlight(token.text, token.lang, function(err, code) {
+          if (err) return done(err);
+          if (code == null || code === token.text) {
+            return --pending || done();
+          }
+          token.text = code;
+          token.escaped = true;
+          --pending || done();
+        });
+      })(tokens[i]);
+    }
+
+    return;
+  }
```

Comment:
This function is almost 100 lines long.  I highly recommend we break it down into pieces.

Reply:
Tagging #1043 so UziTech can get a better feel on internal pull for the updates to the test harness request. No pressure. Just informational purposes.

Reply:
Meant #1103